p2p/host/basic/basic_host.go

```diff
@@ -288,14 +293,32 @@ func (h *BasicHost) newStream(ctx context.Context, p peer.ID, pid protocol.ID) (
 }
 
 // Connect ensures there is a connection between this host and the peer with
-// given peer.ID. Connect will absorb the addresses in pi into its internal
-// peerstore. If there is not an active connection, Connect will issue a
-// h.Network.Dial, and block until a connection is open, or an error is
-// returned.
+// given peer.ID. If there is not an active connection, Connect will issue a
+// h.Network.Dial, and block until a connection is open, or an error is returned.
+// Connect will absorb the addresses in pi into its internal peerstore.
+// It will also resolve any /dns4, /dns6, and /dnsaddr addresses.
 func (h *BasicHost) Connect(ctx context.Context, pi pstore.PeerInfo) error {
+	// TODO(lgierth) figure out whether /ipfs/$peerid needs to be stripped
+	addrs := []ma.Multiaddr{}
+	for _, addr := range pi.Addrs {
+		addrs = append(addrs, addr)
+		p2paddr := ma.Join(addr, ma.StringCast("/ipfs/"+pi.ID.Pretty()))
```

Comment:
There's no nice way (i think). I use ma.Encapsulate normally. Also, prefer to use `IDB58Encode` to `Pretty` because there's no compromise that `Pretty` will always produce the right output.